driver-sync/src/main/com/mongodb/client/internal/ClientSessionBinding.java

```diff
@@ -131,23 +123,23 @@ public OperationContext getOperationContext() {
         return wrapped.getOperationContext();
     }
 
-    private boolean isConnectionSourcePinningRequired() {
-        ClusterType clusterType = wrapped.getCluster().getDescription().getType();
-        return session.hasActiveTransaction() && (clusterType == ClusterType.SHARDED || clusterType == LOAD_BALANCED);
-    }
+    private ConnectionSource getConnectionSource(final Supplier<ConnectionSource> wrappedConnectionSourceSupplier) {
+        if (!session.hasActiveTransaction()) {
+            return wrappedConnectionSourceSupplier.get();
+        }
 
-    private ConnectionSource getPinnedConnectionSource(final boolean isRead) {
-        TransactionContext<Connection> transactionContext = TransactionContext.get(session);
-        ConnectionSource source;
-        if (transactionContext == null) {
-            source = isRead ? wrapped.getReadConnectionSource() : wrapped.getWriteConnectionSource();
-            transactionContext = new TransactionContext<>(wrapped.getCluster().getDescription().getType());
-            session.setTransactionContext(source.getServerDescription().getAddress(), transactionContext);
-            transactionContext.release();  // The session is responsible for retaining a reference to the context
+        if (TransactionContext.get(session) == null) {
+            ConnectionSource source = wrappedConnectionSourceSupplier.get();
+            ClusterType clusterType = source.getServerDescription().getClusterType();
+            if (clusterType == SHARDED || clusterType == LOAD_BALANCED) {
```

Comment:
This is the crux of the refactoring: we obtain the cluster type from the server description, avoiding the need to call `Cluster#getDescription` prior to server selection.